haystack/preview/components/embedders/openai_document_embedder.py

```diff
@@ -0,0 +1,157 @@
+from typing import List, Optional, Dict, Any
+
+import openai
+from tqdm import tqdm
+
+
+from haystack.preview import component, Document, default_to_dict, default_from_dict
+
+
+API_BASE_URL = "https://api.openai.com/v1"
+
+
+@component
+class OpenAIDocumentEmbedder:
+    """
+    A component for computing Document embeddings using OpenAI models.
+    The embedding of each Document is stored in the `embedding` field of the Document.
+    """
+
+    def __init__(
+        self,
+        api_key: str,
```

Comment:
I see this is discussed in more detail here: https://github.com/deepset-ai/haystack/pull/5801#discussion_r1327301037

Reply:
That discussion is unrelated, it talks about serialisation of secret keys.

What you're suggesting is feasible and is not affect by the other discussion.

Reply:
Sure, that part I understand. 🙂 The default value is not related to the serialization. Why I think dealing with `api_key` might still be related is because if the `api_key` here was never really an api key but just the name of an env variable then we wouldn't have a problem serializing it. Maybe the user doesn't even need a way to pass an api key directly. We can encourage them to always pass just the name of an env variable.

Reply:
`api_key` must not be an env var key, that's just wrong.

The parameter is called `api_key` and its value must a valid API key, we can't reliably differentiate from a string that is and env var key or an actual API key.

Making the `api_key` optional and search for the key in the env vars when `api_key` is not provided works though. But we must not make the env var key customizable.

Reply:
We are talking about 2 different topics:
- `api_key` is optional or not?
- serialization and secrets handling

In any case, when #5801 will be in a good shape regarding these topics,
I will apply the same decision to this PR.

Reply:
Addressed these points based on #5801.
- `api_key` can be provided explicitly or read from the env variable `OPENAI_API_KEY`
Please note that there are valid cases where passing it explicitly is needed: https://github.com/deepset-ai/haystack/pull/5801#discussion_r1327545831
- `api_key` is never serialized

Reply:
Sounds great!